Improve RobotInteraction thread safety

acornacorn opened the pull request:
This change attempts to simplify RobotInteraction and improve thread safety.

These changes focus on InteractionHandler.  Some of the functionality is pulled out into helper classes (LockedRobotState and KinematicOptionsMap) so that the intent of InteractionHandler is more clear.

All non-const members are now protected with internal locks.
The locking scheme is documented in comments.

The state update is simplified, removing the need for condition variables.  Instead in-place updates are done using callback functions with the modifyState method (inherited by RobotInteraction from LockedRobotState).

Commit:
robot_interaction: use KinematicOptionsMap

Fixes threading issues.
Separate the handling of kinematics options into a separate object which
enforces thread safe access.

Commit:
pull RobotInteraction structures out of class

The Generic, EndEffector, and Joint structures complicate the core of
RobotInteraction.  Pull them out to simplify the code.  This will also
help with future plans to make the core of RobotInteraction more
generic and flexible.

Commit:
robot_interaction: add KinematicOptions

KinematicOptions contains the parameters needed to call RobotState::setFromIK.

KinematicOptionsMap contains a map of string->KinematicOptions a default KinematicOptions.

These are useful in RobotInteraction with the group name as the key.

Commit:
robot_interaction: add LockedRobotState and tests

Commit:
robot_interaction: use LockedRobotState

Fix a number of thread safety violations.

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction.h:
```diff
@@ -0,0 +1,186 @@
+
+/// Type of function for constructing markers.
+/// This callback sets up the marker used for an interaction.
+typedef boost::function<bool(
```
What does the callback return? How are the arguments used?

acornacorn replied:
Good point -- will add comments

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction_handler.h:
```diff
@@ -37,11 +37,43 @@
 {
 
+class InteractionHandler;
+typedef boost::shared_ptr<InteractionHandler> InteractionHandlerPtr;
```
why not use MOVEIT_CLASS_FORWARD from macros/?

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction.h:
```diff
@@ -0,0 +1,186 @@
+/// Return false if the state was not successfully updated.
+typedef boost::function<bool(
+          robot_state::RobotState&,
+          const visualization_msgs::InteractiveMarkerFeedbackConstPtr &)>
```
Why take a ConstPtr& here for a ROS message and just a & for the callback above? I'd rather we take the ref to the msg in both cases.

acornacorn replied:
This is the existing prototype.  I agree 

```
const visualization_msgs::InteractiveMarkerFeedback &
```

would be better, but changing it would break legacy code.

I plan to change this later by replacing these callback functions with virtual methods in the GenericInteraction class, so this will be improved later.

isucan replied:
That sounds fine. Maybe add a todo? This seems fairly low impact though.. some small things we could maybe break, if they are low level enough that they are unlikely to have been used and it is easy to port code forward.

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction.h:
```diff
@@ -0,0 +1,186 @@
+  InteractionStyle::InteractionStyle interaction;
+
+  /// The size of the end effector group (diameter of enclosing sphere)
+  double size;
```
Should we have a default constructor that sets `size` and `interaction`?

acornacorn replied:
Probably eventually.  This is how it was and I'm inclined to leave it as it until the next phase of changes.

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction.h:
```diff
@@ -0,0 +1,186 @@
+  unsigned int dof;
+
+  /// The size of the connecting link  (diameter of enclosing sphere)
+  double size;
```
Constructor for dof and size?

acornacorn replied:
As above, I'm inclined to leave as is for now.  Will fix later.

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction.h:
```diff
@@ -0,0 +1,186 @@
+
+/// Representation of an interaction via an end-effector
+/// Displays one marker for manipulating the EEF position.
+class EndEffectorInteraction
```
In general the convention in MoveIt is to use structs when all members are public. The object above is a struct. Could we make this one and below structs as well?

acornacorn replied:
will fix.  Eventually I think these may become classes but for now you are right.

isucan commented on robot_interaction/include/moveit/robot_interaction/locked_robot_state.h:
```diff
@@ -0,0 +1,116 @@
+{
+public:
+  LockedRobotState(const robot_state::RobotStatePtr& state);
+  LockedRobotState(const robot_state::RobotState& state);
```
Do we need all these constructors? If not, I propose removing one of the two above. My concern is that the semantics is slightly different. If I use one, then I modify the state I passed to LockedRobotState, for the first case, the change is visible in this object, and for the second, the change is not visible. I would prefer to keep just the plain & constructor, but not the RobotStatePtr& constructor.

acornacorn replied:
Good point.  Will fix (just keep latter 2)

isucan commented on robot_interaction/include/moveit/robot_interaction/locked_robot_state.h:
```diff
@@ -0,0 +1,116 @@
+  // The \e modify function is passed a reference to the state which it can
+  // modify.  No threads will be given access to the state while the \e modify
+  // function is running.
+  void modifyState(ModifyStateFunction modify);
```
can the argument be const &?

acornacorn replied:
fixed

isucan commented on robot_interaction/include/moveit/robot_interaction/locked_robot_state.h:
```diff
@@ -0,0 +1,116 @@
+  // Default definition does nothing.  Override to get notification of state
+  // change.
+  // TODO: is this needed?
+  virtual void robotStateChanged();
```
Can't hurt. I'd say keep it.

isucan commented on robot_interaction/include/moveit/robot_interaction/locked_robot_state.h:
```diff
@@ -0,0 +1,116 @@
+  // The lock can also be used by subclasses to lock additional fields.
+  mutable boost::mutex state_lock_;
+
+private:
```
I think state_ should be protected as well.

acornacorn replied:
I strongly disagree.  The whole purpose of this class is to ensure that state_ is not modified unless (A) it is uniqe and (B) it is done with state_lock_ held.  Making state_ protected will allow violation of this too easily.  If you want to modify the state from the parent class use modifyState().

isucan replied:
Ok, I buy your argument, but it strikes me as odd having a variable and the mutex that protects it at different visibility levels. Can we make that private as well?

acornacorn replied:
We could. But it is useful to share that lock with the subclass so it can use it to lock other members (as mentioned in the comment).  The InteractionHandler (which inherits from this class) uses state_lock_ to protect several of its members.  If InteractionHandler needed its own separate mutex for that it would just mean we have to lock 2 mutexes many places where it is used.

One option is to make state_lock_ private and have a protected member function that grants access to it

```
    boost::mutex getStateLock() const { return state_lock_; }
```

but this seems gratuitous.  It does not really add any protection.

isucan replied:
Ok, I am convinced the current way of doing things makes sense. Thanks for the explanations!

isucan commented on robot_interaction/include/moveit/robot_interaction/robot_interaction.h:
```diff
@@ -39,12 +39,12 @@
-#include <interactive_markers/menu_handler.h>
+//#include <interactive_markers/menu_handler.h>
 #include <moveit/robot_state/robot_state.h>
-#include <moveit/macros/class_forward.h>
```
You seem to not like this... Any reason why? :)

isucan commented on robot_interaction/include/moveit/robot_interaction/robot_interaction.h:
```diff
@@ -211,38 +120,51 @@ class RobotInteraction
+
+  // Get the kinematic options map.
+  // Use this to set kinematic options (defaults or per-group).
+  KinematicOptionsMapPtr getKinematicOptionsMap() { return kinematic_options_map_; }
```
can we return const&?

acornacorn replied:
Since the lifetime of RobotInteraction() is ambiguous it is not clear to me that doing so is safe.  Consider

```
lock_access_to robot_interaction
const KinematicOptionsMapPtr& kinematic_options_map = getKinematicOptionsMap();
unlock_access_to robot_interaction
// some other thread deletes robot_interaction
kinematic_options_map->dereference
```

Am I being too paranoid?  Tell me if this is silly and I'll fix it...

isucan replied:
Technically, you are correct, this can happen. If the user does this though, it is unlikely changing the options map makes any sense. I would allow this to happen -- fairly easy to track down, if it does happen. Your call.

acornacorn replied:
I think you are right that this is unlikely to be a problem.  However I think it makes sense to leave this as is.  This function will really only be used when the options are being modified.  This typically happens from the UI and is not performance critical.  Also all accesses to KinematicOptionsMap require an internal mutex lock and typically an access to a map which mean that the reference inc/dec will not significantly slow things down.

On the other hand if this fails (RobotInteraction is deleted while UI is modifying the value from a different thread)  it will result in a rare race-condition based memory corruption which will be almost impossible to find.

Note: This function is likely to change in my next batch of fixes where I address thread safety in the RobotInteraction class.  We can revisit then, but to be thread safe I think this function will need to return the pointer and not a reference to the pointer.

isucan replied:
Sounds good.

isucan commented on robot_interaction/include/moveit/robot_interaction/robot_interaction.h:
```diff
@@ -211,38 +120,51 @@ class RobotInteraction
-  static bool updateState(robot_state::RobotState &state, const Joint &vj, const geometry_msgs::Pose &pose);
+
+
+#if 0
```
Maybe just delete the disabled code?

acornacorn replied:
oops -- yes, will fix

isucan commented on robot_interaction/src/interaction_handler.cpp:
```diff
@@ -54,69 +56,108 @@
+{
+}
+
+std::string InteractionHandler::fixName(std::string name)
```
This should probably just be a function in an anonymous namespace.

acornacorn replied:
I think it makes sense as a private static method, but I can change it if you really prefer that.  Just curious why that is better?

isucan replied:
I prefer keeping things at the minimum visibility necessary. If someone looks at the header file, they need not see this function, although static private is fine.

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction_handler.h:
```diff
@@ -315,16 +295,61 @@ class RobotInteraction::InteractionHandler
+  const kinematics::KinematicsQueryOptions& getKinematicsQueryOptions() const;
+  void setKinematicsQueryOptions(const kinematics::KinematicsQueryOptions &opt);
+  void setKinematicsQueryOptionsForGroup(const std::string& group_name, 
+           kinematics::KinematicsQueryOptions &options);
```
const options?

acornacorn replied:
will fix

isucan commented on robot_interaction/src/interaction_handler.cpp:
```diff
@@ -332,47 +319,152 @@ void RobotInteraction::InteractionHandler::handleJoint(const robot_interaction::
+{
+  // This is called with state_lock_ held, so no additional locking needed to
+  // access kinematic_options_map_.
+  KinematicOptions kinematic_options = kinematic_options_map_->getOptions(eef->parent_group);
```
const& for return type?

acornacorn replied:
KinematicOptionsMap locks its members internally and only returns copies of its internal state.  Otherwise the KinematicOptionsMap lock (currently private) would have to be locked (and unlocked) by any function calling KinematicOptionsMap methods.  Using the state_lock_ to protect the contents of kinematic_options_map_ would not work because it is shared with RobotInteraction and all the other InteractionHandlers (and state_lock_ is specific to this InteractionHandler instance).

isucan commented on robot_interaction/src/interaction_handler.cpp:
```diff
@@ -407,4 +499,156 @@ bool RobotInteraction::InteractionHandler::transformFeedbackPose(const visualiza
+  // The old *kinematic_options_map_ is automatically deleted by boost::shared_ptr.
+  //
+  // This is a nop if a constructor with a robot_interaction parameter is used.
+  kinematic_options_map_ = shared_kinematic_options_map;
```
You could actually just call .swap() here.

acornacorn replied:
Is there a reason swapping boost::shared_ptr is better than assigning?

isucan replied:
Actually, I am not sure. I have not thought about this in detail, but theoretically you do not need to increase count / decrease count for either of them -- you just swap content. I think swap() may be slightly faster, but I am not sure.

isucan commented on robot_interaction/include/moveit/robot_interaction/kinematic_options.h:
```diff
@@ -0,0 +1,106 @@
+// functions and no destructor.
+struct KinematicOptions
+{
+public: 
```
No need for `public:` in struct.

acornacorn replied:
Since this struct has methods I wanted to be clear that its all public (since some might see the methods and assume it is a class.  Is that bad?  I can remove it if it bugs you.

isucan replied:
I see your point. I have a mild preference for removing it, but it is your call.

isucan replied:
Or maybe we can simply make this a class?

acornacorn replied:
Fixed:
I removed the public:

isucan commented on robot_interaction/include/moveit/robot_interaction/kinematic_options.h:
```diff
@@ -0,0 +1,106 @@
+  /// @param tip link that will be posed
+  /// @param pose desired pose of tip link
+  /// @param result true if IK succeeded.
+  void setStateFromIK(robot_state::RobotState* state,
```
I am all for pointers, but this is slightly inconsistent with the rest of the components in the system. Perhaps I have not seen all the call sites, but the function internally assumes none of the values are nullptr. I think we should set all but `state` to be const & and result should be a return value of the function.

acornacorn replied:
This is intended to be called through a boost::bind.  Using boost::bind with references requires boost::ref() or boost::cref() to pass by value.  For const parameters forgetting boost::cref() just means inefficiency.  For non-const parameters forgetting boost::ref() would mean incorrect behavior with no compiler warning.  I prefer to leave these all as pointers for this reason.

The return value I could be convinced to change.  But this function is intended to be callable through the LockedRobotState::modifyState() function, and that takes a function object with void return.  I could make that function return a bool, but it was unclear to me what a bool return value should mean.  Does it mean the function failed to set the state?  Or (as when using setStateFromIK) does it mean it succeeded in setting the state but the state is somehow invalid?  

Do you think the LockedRobotState::modifyState() method should have a return value?  If so,  what comment would you use to describe the meaning of the return value (it would be the same as the return value of RobotInteraction::updateState() and GenericInteraction's ProcessFeedbackFn callback -- what comment would you use to describe those return values?)

isucan replied:
I see your point. If that is the intention though, why not make the function protected or private? I strongly agree with the cref / ref business. We really should avoid needing that. I don't think modifyState needs to return a value, but if we can generically define a notion of success vs failure, returning bool would be nice.

acornacorn replied:
I can add a bool return value.  I just don't understand what it is supposed to mean.  It doesn't mean IK failed to find a solution.  The RobotState may still be changed even if it returns false (right?).  I guess my real problem is that I do not understand the meaning of the value returned from RobotState::setFromIK().

The whole purpose of this class is to call this function.  Making it private would defeat the purpose (every user of the struct would have to be a friend).

However, I think I actually do not call this from modifyState -- there is an intermediate function -- so I can probably change this to use references.  Will look at that...

isucan replied:
setFromIK should return true if the function succeeded if the IK call succeeded and we set the state to a valid state, according to the constraints we use (which could be empty). There is a bit of weirdness here in the sense that the state could be destroyed even if we return false. I agree with your points above, but if we can keep the ptr stuff a little more hidden for the stuff that can be const&, I'd be happy.

isucan commented on robot_interaction/src/kinematic_options.cpp:
```diff
@@ -0,0 +1,132 @@
+      const geometry_msgs::Pose* pose,
+      bool* result) const
+{
+  const robot_model::JointModelGroup *jmg = state->getJointModelGroup(*group);
```
We should check that jmg is not nullptr and return false if it is.

acornacorn replied:
good point.  Will fix

isucan commented on robot_interaction/include/moveit/robot_interaction/interaction_handler.h:
```diff
@@ -213,89 +181,101 @@ class RobotInteraction::InteractionHandler
+  /** \brief This should only be called by RobotInteraction.
+   * Associates this InteractionHandler to a RobotInteraction.
+   */
+  void setRobotInteraction(RobotInteraction* robot_interaction);
```
I don't think this should be a public function.

isucan replied:
If you really need to, I think friend functions as a last resort are better than the public function, but I think that we can avoid even that.

acornacorn replied:
I really wanted to make this a private function, but to do so would mean RobotInteraction would have to be a friend of InteractionHandler, which I really don't want to do (breaks encapsulation).  I considered adding a new class which is a friend of InteractionHandler and is inherited by RobotInteraction with a single method that calls setRobotInteraction, but that seemed like a lot of complexity.

On the other hand, making this a public function does not really hurt anything.  It is an error (print error and do nothing) to call it with anything other than the RobotInteraction it is already associated with, and calling it with that RobotInteraction does nothing.  So it is benign to call this.  

I could put it in a separate (but still public) section of the class (e.g. at the end) to make it clear it is not part of the public interface.  Another idea is to name it specially (e.g. leading or trailing underscore) to make it clear it is  not really part of the public interface.

I think the best solution is to have the RobotInteraction be a parameter of the InteractionHandler constructor.  This function will not be needed anymore.  But legacy code may still use the old constructor so I don't know if I can get rid of it (your opinion?), which means this function is still needed.

I struggled with this for way longer than such a trivial detail merits, so if you have any good solutions to this problem in general I'd love to hear them.

isucan replied:
I see. I think this is a class that is very unlikely to be present in user
code, and i think the constructor argument is the best option (slightly
breaking api). Better than using friend, I think.
On Feb 15, 2014 2:10 PM, "Acorn" notifications@github.com wrote:

> In
> robot_interaction/include/moveit/robot_interaction/interaction_handler.h:
> 
> >    void clearError(void);
> > -  /*\* \brief This should only be called by RobotInteraction.
> > -   \* Associates this InteractionHandler to a RobotInteraction.
> > -   */
> > -  void setRobotInteraction(RobotInteraction\* robot_interaction);
> 
> I really wanted to make this a private function, but to do so would mean
> RobotInteraction would have to be a friend of InteractionHandler, which I
> really don't want to do (breaks encapsulation). I considered adding a new
> class which is a friend of InteractionHandler and is inherited by
> RobotInteraction with a single method that calls setRobotInteraction, but
> that seemed like a lot of complexity.
> 
> On the other hand, making this a public function does not really hurt
> anything. It is an error (print error and do nothing) to call it with
> anything other than the RobotInteraction it is already associated with, and
> calling it with that RobotInteraction does nothing. So it is benign to call
> this.
> 
> I could put it in a separate (but still public) section of the class (e.g.
> at the end) to make it clear it is not part of the public interface.
> Another idea is to name it specially (e.g. leading or trailing underscore)
> to make it clear it is not really part of the public interface.
> 
> I think the best solution is to have the RobotInteraction be a parameter
> of the InteractionHandler constructor. This function will not be needed
> anymore. But legacy code may still use the old constructor so I don't know
> if I can get rid of it (your opinion?), which means this function is still
> needed.
> 
> I struggled with this for way longer than such a trivial detail merits, so
> if you have any good solutions to this problem in general I'd love to hear
> them.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/ros-planning/moveit_ros/pull/422/files#r9774684
> .

acornacorn replied:
Sounds good.  I'll fix this in stages.  I've already added the new constructor.  Next (after this is merged) I'll update the places that use this code to use the new constructors.  Then I'll eliminate the old constructors and this function.

isucan commented on robot_interaction/include/moveit/robot_interaction/kinematic_options_map.h:
```diff
@@ -0,0 +1,116 @@
+
+// Maintains a set of KinematicOptions with a key/value mapping and a default
+// value.
+struct KinematicOptionsMap
```
This should be a class.

acornacorn replied:
will fix.

isucan commented on robot_interaction/include/moveit/robot_interaction/kinematic_options_map.h:
```diff
@@ -0,0 +1,116 @@
+  /// @param tip link that will be posed
+  /// @param pose desired pose of tip link
+  /// @param result true if IK succeeded.
+  void setStateFromIK(robot_state::RobotState* state,
```
Same comment as previously. key, group, tip, pose should be const&. There is no need to state to be ptr either (for consistency reasons), although I do like having a ptr so it is apparent to the caller the arg is modified. Result should be returned by value. Please please please make this change :) Pretty please.

acornacorn replied:
See reply to other comment

isucan commented on robot_interaction/src/interaction_handler.cpp:
```diff
@@ -47,6 +48,7 @@
 #include <boost/math/constants/constants.hpp>
 #include <algorithm>
 #include <limits>
+#include <stdexcept>
```
Why do we include this if we don't throw exceptions? Also, we should probably use moveit/exceptions/

acornacorn replied:
oops.  will fix

isucan commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -614,8 +641,10 @@ bool RobotInteraction::updateState(robot_state::RobotState &state, const Joint &
 
-bool RobotInteraction::updateState(robot_state::RobotState &state, const EndEffector &eef, const geometry_msgs::Pose &pose,
+// TODO: can we get rid of this?  Only used in moveit_ros/benchmarks_gui/src/tab_states_and_goals.cpp right now.
+bool RobotInteraction::updateState(robot_state::RobotState &state, const EndEffectorInteraction &eef, const geometry_msgs::Pose &pose,
```
+1 for getting rid of it.

isucan commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -178,7 +185,7 @@ void RobotInteraction::decideActiveJoints(const std::string &group)
         if (vj[i].type_ == "planar" || vj[i].type_ == "floating")
         {
-          Joint v;
+          JointInteraction v;
```
+100 for these renames! Thank you!

acornacorn replied:
:)

isucan commented on robot_interaction/src/kinematic_options.cpp:
```diff
@@ -0,0 +1,132 @@
+
+  // This needs to represent all the fields in
+  // robot_interaction::KinematicOptions except options_
+  #define O_FIELDS(F) \
```
Normally I'd love running this in my head to see what is going on but I wanna watch a movie
+1 for coolness :)

acornacorn replied:
thanks :)

davetcoleman replied:
+1 for using "coolness" as a vote

isucan commented on robot_interaction/src/kinematic_options_map.cpp:
```diff
@@ -0,0 +1,154 @@
+  // Lock the one with lower address first.
+  boost::mutex *m1 = &lock_;
+  boost::mutex *m2 = &other.lock_;
+  if (m2 < m1)
```
cool!

isucan commented on robot_interaction/src/kinematic_options_map.cpp:
```diff
@@ -0,0 +1,154 @@
+  if (&other == this)
+    return;
+
+  // need to lock in consitent order to avoid deadlock. 
```
typo

acornacorn replied:
will fix

isucan commented:
+1 for this change. There are some comments that I'd really like to see addressed (setStateFromIK, for instance) and I had some questions in a few places, but this is a very cool improvement!  Thanks!

Commit:
robot_interaction: Fix issues raised by Ioan

Commit:
add missing headers

acornacorn commented:
I've addressed Ioan's comments (all the ones I marked "will fix")

Commit:
remove extraneous code

Commit:
robot_interaction: fix header problems

fix getRobotModel() bug
make internal functions private.

Commit:
robot_interaction: fix setStateFromIK prototypes

use references instead of pointers.

Commit:
robot_interaction: fix comment formatting

Limit lines to 120 chars max (80 preferred in headers).

Commit:
robot_interaction: fix formatting

remove tabs and whitespace at the end of lines.

Commit:
robot_interaction: add comments

Comment cryptic public function behavior.

acornacorn commented:
I've fixed the setStateFromIK() prototypes (use references instead of pointers as requested by Ioan).
Also updated some comments and formatting.

I believe all Ioan's suggestions have now been addressed.
